akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -39,92 +118,88 @@ import scala.util.control.NonFatal
 
-  // a counter that gets incremented every time a task is queued
-  private val executionCounter: AtomicInteger = new AtomicInteger(0)
   // maps a runnable to an index of a worker queue
-  private val runnableToWorkerQueueIndex = new AtomicReference(ImmutableIntMap.empty)
-
-  private def locked[T](l: Lock)(body: ⇒ T) = {
-    l.lock()
-    try {
-      body
-    } finally {
-      l.unlock()
-    }
-  }
-
-  private def getQueueForRunnable(command: Runnable) = {
+  private[this] final val hashCache = new AtomicReference(ImmutableIntMap.empty)
 
+  private def getQueueForRunnable(command: Runnable): BoundedAffinityTaskQueue = {
     val runnableHash = command.hashCode()
 
-    def sbhash(i: Int) = reverseBytes(i * 0x9e3775cd) * 0x9e3775cd
-
-    def getNext = executionCounter.incrementAndGet() % parallelism
+    def indexFor(h: Int): Int =
+      Math.abs(reverseBytes(h * 0x9e3775cd) * 0x9e3775cd) % parallelism // In memory of Phil Bagwell
 
-    def updateIfAbsentAndGetQueueIndex(
-      workerQueueIndex: AtomicReference[ImmutableIntMap],
-      runnableHash:     Int, queueIndex: ⇒ Int): Int = {
-      @tailrec
-      def updateIndex(): Unit = {
-        val prev = workerQueueIndex.get()
-        if (!runnableToWorkerQueueIndex.compareAndSet(prev, prev.updateIfAbsent(runnableHash, queueIndex))) {
-          updateIndex()
+    val workQueueIndex =
+      if (fairDistributionThreshold == 0)
+        indexFor(runnableHash)
+      else {
+        @tailrec
+        def updateAndOrGetIndex(): Int = {
+          val prev = hashCache.get()
+          if (prev.size > fairDistributionThreshold) indexFor(runnableHash)
+          else {
+            val existingIndex = prev.get(runnableHash)
+            if (existingIndex >= 0) existingIndex
+            else {
+              val index = prev.size % parallelism
+              if (hashCache.compareAndSet(prev, prev.updated(runnableHash, index)))
+                index // Successfully added key
+              else
+                updateAndOrGetIndex() // Try again
+            }
+          }
         }
-      }
-      updateIndex()
-      workerQueueIndex.get().get(runnableHash) // can safely call get..
-    }
 
-    val workQueueIndex =
-      if (fairDistributionThreshold == 0 || runnableToWorkerQueueIndex.get().size > fairDistributionThreshold)
-        Math.abs(sbhash(runnableHash)) % parallelism
-      else
-        updateIfAbsentAndGetQueueIndex(runnableToWorkerQueueIndex, runnableHash, getNext)
+        updateAndOrGetIndex()
+      }
 
     workQueues(workQueueIndex)
   }
 
-  //fires up initial workers
-  locked(bookKeepingLock) {
-    workQueues.foreach(q ⇒ addWorker(workers, q))
-  }
-
-  private def addWorker(workers: mutable.Set[ThreadPoolWorker], q: BoundedTaskQueue): Unit = {
-    locked(bookKeepingLock) {
-      val worker = new ThreadPoolWorker(q, new IdleStrategy(idleCpuLevel))
-      workers.add(worker)
-      worker.startWorker()
+  def start(): this.type =
```

Comment:
@zaharidichev I changed this to be an externally called method in order to not publish `this` to the workers before the pool is fully instantiated.